code/datums/spells/spacetime_dist.dm

```diff
@@ -0,0 +1,160 @@
+	return ..() && ready
+
+/obj/effect/proc_holder/spell/spacetime_dist/cast(list/targets, mob/user = usr)
+	. = ..()
+	var/list/turf/to_switcharoo = get_targets_to_scramble(usr)
+	if(!length(to_switcharoo))
+		to_chat(usr, "<span class='warning'>For whatever reason, the strings nearby aren't keen on being tangled.</span>")
+		return
+
+	ready = FALSE
+
+	for(var/turf/swap_a as anything in to_switcharoo)
+		var/turf/swap_b = to_switcharoo[swap_a]
+		var/obj/effect/cross_action/spacetime_dist/effect_a = new /obj/effect/cross_action/spacetime_dist(swap_a)
+		var/obj/effect/cross_action/spacetime_dist/effect_b = new /obj/effect/cross_action/spacetime_dist(swap_b)
+		effect_a.linked_dist = effect_b
+		effect_a.add_overlay(swap_b.photograph())
+		effect_b.linked_dist = effect_a
+		effect_b.add_overlay(swap_a.photograph())
+		effect_b.set_light(4, 30, "#c9fff5")
+		LAZYADD(effects, effect_a)
+		LAZYADD(effects, effect_b)
+
+/turf/proc/photograph(limit = 20)
+	var/image/I = new()
+	I.add_overlay(src)
+	for(var/V in contents)
+		var/atom/A = V
+		if(A.invisibility)
+			continue
+		I.add_overlay(A)
+		if(limit)
+			limit--
+		else
+			return I
+	return I
+
+/obj/effect/proc_holder/spell/spacetime_dist/after_cast(list/targets, mob/user)
+	. = ..()
+	addtimer(CALLBACK(src, PROC_REF(clean_turfs)), duration)
+
+/// Callback which cleans up our effects list after the duration expires.
+/obj/effect/proc_holder/spell/spacetime_dist/proc/clean_turfs()
+	QDEL_LIST(effects)
+	ready = TRUE
+
+/**
+ * Gets a list of turfs around the center atom to scramble.
+ *
+ * Returns an assoc list of [turf] to [turf]. These pairs are what turfs are
+ * swapped between one another when the cast is done.
+ */
+/obj/effect/proc_holder/spell/spacetime_dist/proc/get_targets_to_scramble(atom/center)
+	// Get turfs around the center
+	var/list/turfs = spiral_range_turfs(scramble_radius + 3 * spell_level, center) // 7 10 13 16
+	if(!length(turfs))
+		return
+
+	var/list/turf_steps = list()
+
+	// Go through the turfs we got and pair them up
+	// This is where we determine what to swap where
+	var/num_to_scramble = round(length(turfs) * 0.5)
+	for(var/i in 1 to num_to_scramble)
+		turf_steps[pick_n_take(turfs)] = pick_n_take(turfs)
+
+	// If there's any turfs unlinked with a friend,
+	// just randomly swap it with any turf in the area
+	if(length(turfs))
+		var/turf/loner = pick(turfs)
+		var/area/caster_area = get_area(center)
+		turf_steps[loner] = get_turf(pick(caster_area.contents))
+
+	return turf_steps
+
+
+/obj/effect/cross_action
+	name = "cross me"
+	desc = "for crossing"
+	anchored = TRUE
+
+/obj/effect/cross_action/spacetime_dist
+	name = "spacetime distortion"
+	desc = "A distortion in spacetime. You can hear faint music..."
+	icon_state = "nothing"
+	/// A flags which save people from being thrown about
+	var/obj/effect/cross_action/spacetime_dist/linked_dist
+	var/busy = FALSE
```

Comment:
The var isn't really describing what it does. `can_teleport` or something would probably be more fitting. Also a comment saying it is there to avoid an infinite teleport back and forth loop would be good

Reply:
done